File: app/src/main/java/com/persona5dex/Persona5Application.java
package com.persona5dex;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import androidx.core.app.JobIntentService;

import com.persona5dex.dagger.application.ApplicationContextModule;
import com.persona5dex.dagger.application.DaggerPersona5ApplicationComponent;
import com.persona5dex.dagger.application.Persona5ApplicationComponent;
import com.persona5dex.models.room.PersonaDatabase;

/**
 * Created by Rechee on 7/1/2017.
 */

public class Persona5Application extends Application {
    private Persona5ApplicationComponent component;
    private PersonaDatabase database;

    public static Persona5Application get(Activity activity){
        return (Persona5Application) activity.getApplication();
    }

    public static Persona5Application get(JobIntentService service){
        return (Persona5Application) service.getApplication();
    }

    @Override
    public void onCreate() {
        super.onCreate();

        this.component = DaggerPersona5ApplicationComponent.builder()
                .applicationContextModule(new ApplicationContextModule(this))
                .build();
    }

    public static PersonaDatabase getPersonaDatabase(Context context) {
        Persona5Application application = (Persona5Application) context.getApplicationContext();
        return application.getPersonaDatabase();
    }

    public PersonaDatabase getPersonaDatabase() {
        if(this.database == null) {
            this.database = PersonaDatabase.Companion.getPersonaDatabase(this);
        }

        return database;
    }

    public Persona5ApplicationComponent getComponent() {
        return this.component;
    }
}

File: app/src/main/java/com/persona5dex/Persona5Application.kt
package com.persona5dex

import android.app.Activity
import android.app.Application
import androidx.core.app.JobIntentService
import androidx.work.Configuration
import com.persona5dex.dagger.application.ApplicationContextModule
import com.persona5dex.dagger.application.DaggerPersona5ApplicationComponent
import com.persona5dex.dagger.application.Persona5ApplicationComponent
import com.persona5dex.models.room.PersonaDatabase


/**
 * Created by Rechee on 7/1/2017.
 */

open class Persona5Application : Application(), Configuration.Provider {
    lateinit var component: Persona5ApplicationComponent
        private set
    val database: PersonaDatabase by lazy { PersonaDatabase.getPersonaDatabase(this) }

    override fun onCreate() {
        super.onCreate()

        component = DaggerPersona5ApplicationComponent.builder()
            .applicationContextModule(ApplicationContextModule(this))
            .build()

        component.personaJobCreator().scheduleGenerateFusionJob()
    }

    override fun getWorkManagerConfiguration(): Configuration =
        Configuration.Builder().build()

    companion object {

        @JvmStatic
        fun get(activity: Activity): Persona5Application =
            activity.application as Persona5Application

        @JvmStatic
        fun get(service: JobIntentService): Persona5Application =
            service.application as Persona5Application
    }
}

